src/amltk/sklearn/evaluation.py

```diff
@@ -557,60 +599,99 @@ def cross_validate_task(  # noqa: D103, C901, PLR0912, PLR0915
-            if store_models:
-                trial.store({f"model_{i}.pkl": _trained_estimator})
-
-            trial.summary.update({f"fold_{i}:{k}": v for k, v in _val_scores.items()})
-            for k, v in _val_scores.items():
-                all_val_scores[k].append(v)
-
-            if _train_scores is not None:
+    with comm.open() if comm is not None else nullcontext():
+        try:
+            # Open up comms if passed in, allowing for the cv early stopping mechanism
+            # to communicate back to the main process
+            # Main cv loop
+            for i, (_trained_est, _val_scores, _train_scores) in trial.profiler.each(
+                enumerate(cv_iter),
+                name="cv",
+                itr_name="fold",
+            ):
+                # Update the report
+                if store_models:
+                    trial.store({f"model_{i}.pkl": _trained_est})
+
+                fold_scores = {f"fold_{i}:{k}": v for k, v in _val_scores.items()}
+
+                trial.summary.update(fold_scores)
+                for k, v in _val_scores.items():
+                    all_val_scores[k].append(v)
+
+                if _train_scores is not None:
+                    train_scores = {
+                        f"fold_{i}:train_{k}": v for k, v in _train_scores.items()
+                    }
+                    trial.summary.update(train_scores)
+                    for k, v in _train_scores.items():
+                        all_train_scores[k].append(v)
+                else:
+                    train_scores = None
+
+                # If there was a comm passed, we are operating under cv early stopping
+                # mode, in which case we request information from the main process,
+                # should we continue or stop?
+                if comm is not None and i < n_splits:
+                    rqst_info = CVEvaluation.FoldInfo(
+                        trial=trial,
+                        pipeline=configured_pipeline,
+                        fold=i,
+                        scores=all_val_scores,
+                        train_scores=all_train_scores,
+                    )
+                    match response := comm.request(rqst_info, timeout=10):
+                        case True:
+                            raise CVEarlyStoppedError("Early stopped!")
+                        case False:
+                            pass
+                        case np.bool_() if bool(response) is True:
+                            raise CVEarlyStoppedError("Early stopped!")
+                        case Exception():
+                            raise response
+                        case _:
+                            raise RuntimeError(
+                                f"Recieved {response=} which we can't handle."
+                                " Please return `True`, `False` or an `Exception`"
+                                f" and not a type {type(response)=}",
+                            )
```

Comment:
@LennartPurucker this is the core new logic. After each fold, if we have a `Comm`, we `request(...)` from the main process what we do next.